python/parquet_tools/common/schema.py

```diff
@@ -0,0 +1,700 @@
+          _colSpecs list and pandas type is entered into the ColTypes
+          dictionary with cname as the key.
+        - If the column may be null, the [column number, cname_ISNULL] is
+          appended to the colnulls list passed as an argument. The type for
+          this additional column is also added to the colTypes dictionary.
+        """
+
+        # Check for types using lower case but use original spec in messages
+        #
+        xtype = ctype.lower()
+
+        # Check for character type and decimal type
+        #
+        if xtype[0:4] == 'char' or xtype[0:7] == 'varchar':
+            xtype = 'char'
+        elif xtype[0:8] == 'decimal(':
+            xtype = Schema._get_dectype(xtype[8:], cname)
+        else:
+            # Record columns that have float or double types.
+            #
+            try:
+                if TypeInfo.table[xtype] == np.float32:
+                    Schema._colFP32.append(cname)
+                elif TypeInfo.table[xtype] == np.float64:
+                    Schema._colFP64.append(cname)
+            except KeyError:
+                raise ParmError(4, cname, ctype)
+
+        # All types must be found in our name to numpy dictionary. We have
+        # screened for ths but just on case we check again.
+        #
+        try:
+            nptype = TypeInfo.table[xtype]
+            Schema.colTypes[cname] = nptype
+            Schema._colSpecs.append([cname, ctype, str(np.dtype(nptype))])
+        except KeyError:
+            raise ParmError(4, cname, ctype)
+
+        # Now record whether we should preprocess null values for the column
+        #
+        if cnull and xtype in TypeInfo.isInt:
+            cname += Schema._nilSuffx
+            colnulls.append([len(Schema.colNames)-1, cname])
+            Schema.colTypes[cname] = np.bool_
+
+    # *************************************************************************
+    # *                          c h k _ s c h e m a                          *
+    # *************************************************************************
+    @classmethod
+    def chk_schema(cls, infile, sepc):
+        """Check that the schema coresponds to the data.
+
+        Parameters
+        ----------
+        infile : 'string'
+            The path to the input csv file.
+        sepc : 'string'
+            The csv column separator character.
+
+        Raises
+        -------
+        FatalError
+        ParmError
+        """
+
+        # Screen for the common error here.
+        #
+        if not os.path.exists(infile):
+            raise ParmError(6, 'Schema file', infile, 'not found')
+
+        # Read the first line of the cvs file, which may be the header
+        #
+        try:
+            with open(infile) as f:
+                csvrdr = csv.reader(f, delimiter=sepc, quoting=csv.QUOTE_NONE)
+                row = next(csvrdr)
+                if len(row) != len(Schema._colSpecs):
```

Comment:
see, all this classmethod nonsense makes it look like you're accessing the private data of a different class here.